### ISSUE TITLE
vcpkg & glib impasse 

### ISSUE THREAD

**Issue by interloper98** [dosbox-staging/dosbox-staging#4010]
According to the vcpkg team, our glib 2.78.4 isn't compatible with the newer version of vcpkg's embedded python 3.12.

Our options are rolling back to an older vcpkg that uses python 3.11 or using the newer glib 2.80 (which is compatible with the latest vcpkg tools)

See https://github.com/microsoft/vcpkg/issues/41791#event-14908038274

I grepped for `2.78.4` to find where this glib version is defined, but I'm not getting any hits; how is this stuff defined?

Hunch is there's near zero risk with either approach, but I don't know how to progress it. Need help 📯

**Comment by shermp** [comment]
Package versions are determined by the `builtin-baseline` key in `vcpkg.json`.

glib 2.80.x was added to vcpkg approx 2 weeks ago, which is after our current `builtin-baseline`.

The question is, where does Python come from? Is it the Python pre-installed on the github runners? Or does vcpkg download it from msys2 as part of the glib build?

**Comment by shermp** [comment]
```
 C:\vcpkg\buildtrees\versioning_\versions\glib\d73077209c5fcb5a0340baf7d8ba76d128c1ab27: info: installing overlay port from here
         -- Downloading https://github.com/mesonbuild/meson/archive/614d436232d3a86518164cbe2b8af12db3bde009.tar.gz -> meson-614d436232d3a86518164cbe2b8af12db3bde009.tar.gz...
         -- Applying patch D:/a/dosbox-staging/dosbox-staging/vcpkg_installed/x64-windows/x64-windows/share/vcpkg-tool-meson/meson-intl.patch
         -- Applying patch D:/a/dosbox-staging/dosbox-staging/vcpkg_installed/x64-windows/x64-windows/share/vcpkg-tool-meson/adjust-python-dep.patch
         -- Applying patch D:/a/dosbox-staging/dosbox-staging/vcpkg_installed/x64-windows/x64-windows/share/vcpkg-tool-meson/adjust-args.patch
         -- Applying patch D:/a/dosbox-staging/dosbox-staging/vcpkg_installed/x64-windows/x64-windows/share/vcpkg-tool-meson/remove-freebsd-pcfile-specialization.patch
         -- Applying patch D:/a/dosbox-staging/dosbox-staging/vcpkg_installed/x64-windows/x64-windows/share/vcpkg-tool-meson/meson-1.4-llvm-18.diff
         -- Downloading https://www.python.org/ftp/python/3.12.7/python-3.12.7-embed-amd64.zip -> python-3.12.7-embed-amd64.zip...
         -- Found Python version '3.12.7 at C:/vcpkg/downloads/tools/python/python-3.12.7-x64/python.exe'
         -- Using meson: C:/vcpkg/downloads/tools/meson-1.3.2-aa4767/meson.py
         -- Downloading https://download.gnome.org/sources/glib/2.78/glib-2.78.4.tar.xz -> glib-2.78.4.tar.xz...
         -- Extracting source C:/vcpkg/downloads/glib-2.78.4.tar.xz
         -- Applying patch use-libiconv-on-windows.patch
         -- Applying patch libintl.patch
         -- Using source at C:/vcpkg/buildtrees/glib/src/glib-2-322e03f702.clean
```

There's the answer.

**Comment by shermp** [comment]
We've had a couple of these vcpkg issues lately. I think we need to stop relying on the runner provided vcpkg, and download our own.

**Comment by weirddan455** [comment]
> We've had a couple of these vcpkg issues lately. I think we need to stop relying on the runner provided vcpkg, and download our own.

Yeah, we need this crap to stop auto-updating.  That's the reason why we moved off MSYS2 in the first place.  We can't have our CI be randomly failing due to external updates.

**Comment by shermp** [comment]
Maybe give https://github.com/marketplace/actions/run-vcpkg a go?

**Comment by interloper98** [comment]
Tried rolling the worker's vcpkg back to the vcpkg baseline commit, however it has its own different problems:

https://github.com/dosbox-staging/dosbox-staging/actions/runs/11534575277/job/32108743643

```
         A suitable version of 7zip was not found (required v23.1.0).
         Downloading https://github.com/ip7z/7zip/releases/download/23.01/7z2301-extra.7z
         Extracting 7zip...
     2>EXEC : error : Could not fetch 7zr. vcpkg does not have a definition of this tool for this platform. [D:\a\dosbox-staging\dosbox-staging\tests\vs\tests.vcxproj]
         note: updating vcpkg by rerunning bootstrap-vcpkg may resolve this failure.
```

I'll try adjusting the baseline to roughly 2+ weeks ago and hopefully land on more stable ground in vcpkg land.

**Comment by interloper98** [comment]
Using the 2024-09-30 tagged release, and it's made it past that `7zr` error:

```
A suitable version of 7zip was not found (required v24.8.0).
Downloading https://github.com/ip7z/7zip/releases/download/24.08/7z2408-extra.7z
Extracting 7zip...
A suitable version of 7zr was not found (required v24.8.0).
Downloading https://github.com/ip7z/7zip/releases/download/24.08/7zr.exe
```

 let's see how much further it gets :crossed_fingers:

**Comment by interloper98** [comment]
Finally.. glib is passing!

```
Building glib:x64-windows@2.78.4#4...
         C:\vcpkg\buildtrees\versioning_\versions\glib\ee1411dc98162edf7c0e77780050dc8ddc7b3aae: info: installing overlay port from here
         -- Found Python version '3.11.8 at C:/vcpkg/downloads/tools/python/python-3.11.8-x64/python.exe'
         -- Using meson: C:/vcpkg/downloads/tools/meson-1.5.2-dd4692/meson.py
         -- Downloading https://download.gnome.org/sources/glib/2.78/glib-2.78.4.tar.xz -> glib-2.78.4.tar.xz...
         -- Extracting source C:/vcpkg/downloads/glib-2.78.4.tar.xz
         -- Applying patch use-libiconv-on-windows.patch
         -- Applying patch libintl.patch
         -- Using source at C:/vcpkg/buildtrees/glib/src/glib-2-322e03f702.clean
         -- Found external ninja('1.11.0').
         -- Using cached msys2-mingw-w64-x86_64-pkgconf-1~2.2.0-1-any.pkg.tar.zst.
         -- Using cached msys2-msys2-runtime-3.5.3-3-x86_64.pkg.tar.zst.
         -- Using msys root at C:/vcpkg/downloads/tools/msys2/1e74ca60daa10104
         -- Configuring x64-windows-dbg
         -- Getting CMake variables for x64-windows
         -- Configuring x64-windows-dbg done
         -- Configuring x64-windows-rel
         -- Configuring x64-windows-rel done
         -- Package x64-windows-dbg
         -- Package x64-windows-rel
         -- Fixing pkgconfig file: C:/vcpkg/packages/glib_x64-windows/lib/pkgconfig/gio-2.0.pc
         -- Fixing pkgconfig file: C:/vcpkg/packages/glib_x64-windows/lib/pkgconfig/gio-windows-2.0.pc
         -- Fixing pkgconfig file: C:/vcpkg/packages/glib_x64-windows/lib/pkgconfig/glib-2.0.pc
         -- Fixing pkgconfig file: C:/vcpkg/packages/glib_x64-windows/lib/pkgconfig/gmodule-2.0.pc
         -- Fixing pkgconfig file: C:/vcpkg/packages/glib_x64-windows/lib/pkgconfig/gmodule-export-2.0.pc
         -- Fixing pkgconfig file: C:/vcpkg/packages/glib_x64-windows/lib/pkgconfig/gmodule-no-export-2.0.pc
         -- Fixing pkgconfig file: C:/vcpkg/packages/glib_x64-windows/lib/pkgconfig/gobject-2.0.pc
         -- Fixing pkgconfig file: C:/vcpkg/packages/glib_x64-windows/lib/pkgconfig/gthread-2.0.pc
         -- Fixing pkgconfig file: C:/vcpkg/packages/glib_x64-windows/debug/lib/pkgconfig/gio-2.0.pc
         -- Fixing pkgconfig file: C:/vcpkg/packages/glib_x64-windows/debug/lib/pkgconfig/gio-windows-2.0.pc
         -- Fixing pkgconfig file: C:/vcpkg/packages/glib_x64-windows/debug/lib/pkgconfig/glib-2.0.pc
         -- Fixing pkgconfig file: C:/vcpkg/packages/glib_x64-windows/debug/lib/pkgconfig/gmodule-2.0.pc
         -- Fixing pkgconfig file: C:/vcpkg/packages/glib_x64-windows/debug/lib/pkgconfig/gmodule-export-2.0.pc
         -- Fixing pkgconfig file: C:/vcpkg/packages/glib_x64-windows/debug/lib/pkgconfig/gmodule-no-export-2.0.pc
         -- Fixing pkgconfig file: C:/vcpkg/packages/glib_x64-windows/debug/lib/pkgconfig/gobject-2.0.pc
         -- Fixing pkgconfig file: C:/vcpkg/packages/glib_x64-windows/debug/lib/pkgconfig/gthread-2.0.pc
         -- Installing: C:/vcpkg/packages/glib_x64-windows/share/glib/copyright
         -- Performing post-build validation
         Successfully stored glib_x64-windows.zip to https://acghubeus1.actions.githubusercontent.com/RDKYnb9yFkXPfcXINdFWH3mzDaNqlgzrXM4kV8rnx6PJBbLLEF/_apis/artifactcache/caches/112529.
         Stored binaries in 1 destinations in 3.7 s.
         Elapsed time to handle glib:x64-windows: 2.3 min
         glib:x64-windows package ABI: 3fa33184619d1[442](https://github.com/dosbox-staging/dosbox-staging/actions/runs/11534670187/job/32108948930#step:12:443)bd8b605898e002aac621ba7d422b1ee1c6442e5968f3a921
```

**Comment by interloper98** [comment]
Now blowing up on libogg:

```
         Installing 20/27 libogg:x64-windows@1.3.5#1...
         Building libogg:x64-windows@1.3.5#1...
         C:\vcpkg\buildtrees\versioning_\versions\libogg\98e27727c32704393494d32615c25b7b16eb3067: info: installing overlay port from here
         -- Downloading https://gitlab.xiph.org/xiph/ogg/-/archive/v1.3.5/ogg-v1.3.5.tar.gz -> xiph-ogg-v1.3.5.tar.gz...
         [DEBUG] To include the environment variables in debug output, pass --debug-env
         [DEBUG] Trying to load bundleconfig from C:\vcpkg\vcpkg-bundle.json
         [DEBUG] Failed to open: C:\vcpkg\vcpkg-bundle.json
         [DEBUG] Bundle config: readonly=false, usegitregistry=false, embeddedsha=nullopt, deployment=Git, vsversion=nullopt
         [DEBUG] Metrics enabled.
         [DEBUG] Feature flag 'binarycaching' unset
         [DEBUG] Feature flag 'compilertracking' unset
         [DEBUG] Feature flag 'registries' unset
         [DEBUG] Feature flag 'versions' unset
         [DEBUG] Feature flag 'dependencygraph' unset
     3>EXEC : warning : Download failed -- retrying after 1000ms [D:\a\dosbox-staging\dosbox-staging\src\libs\zmbv\zmbv.vcxproj]
     3>EXEC : warning : Download failed -- retrying after 2000ms [D:\a\dosbox-staging\dosbox-staging\src\libs\zmbv\zmbv.vcxproj]
     3>EXEC : warning : Download failed -- retrying after 4000ms [D:\a\dosbox-staging\dosbox-staging\src\libs\zmbv\zmbv.vcxproj]
     3>EXEC : error : Missing xiph-ogg-v1.3.5.tar.gz and downloads are blocked by x-block-origin. [D:\a\dosbox-staging\dosbox-staging\src\libs\zmbv\zmbv.vcxproj]
     3>EXEC : error : https://gitlab.xiph.org/xiph/ogg/-/archive/v1.3.5/ogg-v1.3.5.tar.gz: WinHttpSendRequest failed with exit code 12002 [D:\a\dosbox-staging\dosbox-staging\src\libs\zmbv\zmbv.vcxproj]
     3>EXEC : error : https://gitlab.xiph.org/xiph/ogg/-/archive/v1.3.5/ogg-v1.3.5.tar.gz: WinHttpSendRequest failed with exit code 12002 [D:\a\dosbox-staging\dosbox-staging\src\libs\zmbv\zmbv.vcxproj]
     3>EXEC : error : https://gitlab.xiph.org/xiph/ogg/-/archive/v1.3.5/ogg-v1.3.5.tar.gz: WinHttpSendRequest failed with exit code 12002 [D:\a\dosbox-staging\dosbox-staging\src\libs\zmbv\zmbv.vcxproj]
     3>EXEC : error : https://gitlab.xiph.org/xiph/ogg/-/archive/v1.3.5/ogg-v1.3.5.tar.gz: WinHttpSendRequest failed with exit code 12002 [D:\a\dosbox-staging\dosbox-staging\src\libs\zmbv\zmbv.vcxproj]
         [DEBUG] D:\a\_work\1\s\src\vcpkg\base\downloads.cpp(1030): 
         [DEBUG] Time in subprocesses: 0us
         [DEBUG] Time in parsing JSON: 6us
         [DEBUG] Time in JSON reader: 0us
         [DEBUG] Time in filesystem: 856us
         [DEBUG] Time in loading ports: 0us
         [DEBUG] Exiting after 1.5 min (91248420us)
         
         CMake Error at scripts/cmake/vcpkg_download_distfile.cmake:32 (message):
```

This is a delicate clownshow :clown_face: !  Time to adjust the baseline a bit more (after the ogg fix but before the move to python 3.12)

**Comment by shermp** [comment]
Oh FFS. Is VCPKG really better than MSYS2 at this point?

**Comment by interloper98** [comment]
> Oh FFS. Is VCPKG really better than MSYS2 at this point?

In this latest error, "Missing xiph-ogg-v1.3.5.tar.gz", it tried to download the source package but the only hosting repo at "https://gitlab.xiph.org" is just gone completely (I can't even get to the top level website?).  And VCPKG doesn't have mirrors for it, I guess?

Hopefully this next baseline has better luck.

**Comment by weirddan455** [comment]
> Oh FFS. Is VCPKG really better than MSYS2 at this point?

I don't know.  This is a mess.  We could theoretically setup a Linux runner to do cross compilation for Windows using the mingw packages + the normal Meson build.  I've done that before on other projects and it gives a stable build environment.  Not saying that's the best idea but :shrug:.  I'm getting a bit frustrated with these Windows build tools.

**Comment by interloper98** [comment]
The wheels have come off and I've hit the end of the line.

I went right up to commit just before the Python 3.12 update, but it's still failing out. I've opened [a new ticket](https://github.com/microsoft/vcpkg/issues/41803) for the upstream issue.

Well.. I'm outta tries for now to get 2.78.4 going; I'll try one last shot just with the latest commit ID for the hell of it.

**Comment by shermp** [comment]
What's really annoying is that in theory, all our dependencies should be cached, but vcpkg is really aggressive about cache invalidation on the tiniest of changes.

**Comment by interloper98** [comment]
Unfortunately the gap between a couple weeks ago and the latest commit doesn't improve things for libogg (no commits that add mirrors or fallbacks), so I think we're cooked. Gotta wait this this one gets fixed.

**Comment by shermp** [comment]
Hence, we should probably stop relying on the GH provided vcpkg

**Comment by interloper98** [comment]
I was checking out the old vcpkg version back in time (`c:\vcpkg` is just a git repo itself, it has all the commits that any other cloned copy would have).

I'm not sure how changing to a different vcpkg source would help us here, unless you're talking about us making changes ourself or maybe using a fork or pointing to a different binary cache that doesn't invalidate as frequently?

**Comment by shermp** [comment]
So, in theory (in my mind at least) building a package for a certain baseline should always have a stable hash for binary caching. Unfortunately, this doesn't seem to be the case, especially for `glib`, so there must be something about the vcpkg environment that breaks this.

Although, maybe it's a compiler update that breaks the cache hash? But we are clearly using some cached dependencies even when `glib` needs to be recompiled.

**Comment by interloper98** [comment]
That makes sense; I'm also not sure what all goes into the binary cache or how widely they're shared (presumably if they hashed the compiler itself, flags, and source files, they could binary cache .o files everyone).

A couple other people reported the libogg issue too - and surprisingly their CI workflows are also not benefiting from a binary cache, either.

**Comment by shermp** [comment]
We do binary caching using vcpkg's inbuilt github artifacts caching mechanism. So it's a bit of a black box.

**Comment by johnnovak** [comment]
Guys, if the vcpkg baseline pinning doesn't work as it should, that should be fixed and that's the end of it. We should never ever get updated packages unless we commit a new baseline. My understanding is this is how @kklobe did it and it was working -- maybe someone change it?

Let's not conflate two things either:

- the ability to use a fixed baseline and opt-out of rolling dependencies 100%
- updating the baseline to ther very latest occasionally breaking something

MSYS2 doesn't bring us point one -- that's a showstopper, just as it has always been.

If number two is not true, that must be fixed as we have screwed up. One of the big selling point of vcpkg is the fixed baselines you upgrade at your own pace once th8ngs have stabilised.

We should stay 2-4 weeks or more before the latest and only upgrade at our pace. The build was green a week ago, so just reset the baseline to the one from last week and make sure nothing gets auto updated.

Am I missing something?

**Comment by johnnovak** [comment]
> What's really annoying is that in theory, all our dependencies should be cached, but vcpkg is really aggressive about cache invalidation on the tiniest of changes.

But something doesn't make sense -- we use a fixed baseline, don't we? Cache invalidation or not, we should get the same baseline, no matter what.

**Comment by darkstar** [comment]
At least locally, the MSVC build works fine now by switching to the newest vkpkg baseline e60236 👍

**Comment by shermp** [comment]
I've done some investigation, it seems the `scripts/cmake/vcpkg_find_acquire_program(PYTHON3).cmake` was updated to Python 3.12 last week. This makes me suspect that the scripts are not considered as part of the baseline.

The solution I think is to vendor vcpkg itself. Either as a submodule (yuck), or in the GH action.

**Comment by johnnovak** [comment]
> I've done some investigation, it seems the `scripts/cmake/vcpkg_find_acquire_program(PYTHON3).cmake` was updated to Python 3.12 last week. This makes me suspect that the scripts are not considered as part of the baseline.
> 
> The solution I think is to vendor vcpkg itself. Either as a submodule (yuck), or in the GH action.

Yeah, and that's detail the end user should have zero care about. All I care about as a user a) having the latest vcpkg installed, b) having a baseline hash from potentially years ago. It's then vcpkg's job to make it happen, as per my comment [here](https://github.com/microsoft/vcpkg/issues/41791#issuecomment-2439760998). If it needs to install six different Python runtimes and three Ruby runtimes, so be it—I don't care, just make it work 😆

**Comment by shermp** [comment]
In the case of `glib` this is really ugly, because as far as I can tell, the `glib` port does not actually declare its Python dependency. Python is downloaded because `meson` requires it, and `meson` doesn't really care about Python versions too much.

TLDR the `glib` port is a horrible hack because it implicitly depends on software its build system depends on.

**Comment by johnnovak** [comment]
> In the case of `glib` this is really ugly, because as far as I can tell, the `glib` port does not actually declare its Python dependency. Python is downloaded because `meson` requires it, and `meson` doesn't really care about Python versions too much.
> 
> TLDR the `glib` port is a horrible hack because it implicitly depends on software its build system depends on.
> 
> EDIT: And **** Python for breaking its standard library.

Ok, so glib is a miserable piece of shit—news at 11 😆

Well, the beauty of vcpkg is its rather large user base, so we can lean back and wait for the fix. The other good news is nobody holds a gun to our head to release something tomorrow, _and_ the 0.82.0 final is already built, so we're winning 🥳

**Comment by interloper98** [comment]
Hmmm 🤔 the last attempt passed: https://github.com/dosbox-staging/dosbox-staging/actions/runs/11535119127

Edit: looks like someone rebooted the gitlab.xiph.org box. 

I'll revisit that prior attempt to use glib 2.78.4 + python 3.11.x.

**Comment by johnnovak** [comment]
> Edit: looks like someone rebooted the gitlab.xiph.org box.

Yeah, so to be fair to vcpkg, it's not their fault the external links stop working 😄 On the other hand, with Microsoft's resources mirroring everything going back for years would be a drop in the ocean for them.
